tools/build_langserver/langserver/request_store.go

```diff
@@ -0,0 +1,57 @@
+package langserver
+
+import (
+	"context"
+	"sync"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+type requestStore struct {
+	mu       sync.Mutex
+	requests map[jsonrpc2.ID]request
+}
+
+// Perhaps later we can store more things in the request we might want to use
+type request struct {
+	id     string
+	cancel func()
+}
+
+func (rs *requestStore) Store(ctx context.Context, req *jsonrpc2.Request) context.Context {
+	ctx, cancel := context.WithCancel(ctx)
+
+	rs.mu.Lock()
+	if rs.requests == nil {
```

Comment:
it'd be easier to initialise this at construction time, then you don't have to worry if it's nil.